File: android/app/src/main/java/io/github/chikyukido/scrobblium/database/SongData.java
package io.github.chikyukido.scrobblium.database;


import android.media.MediaMetadata;
import android.media.session.MediaController;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import java.time.LocalDateTime;
import java.util.Objects;

@Entity(tableName = "played_songs")
public class SongData {
    private final String artist;
    private final String title;
    private final String album;
    private final String albumAuthor;
    private final long maxProgress;
    private final LocalDateTime startTime;
    @PrimaryKey(autoGenerate = true)
    private long id;
    private long progress;
    private LocalDateTime endTime;
    private int timeListened;

    public SongData(String artist, String title, String album, String albumAuthor, long progress, long maxProgress, LocalDateTime startTime, LocalDateTime endTime, int timeListened) {
        this.artist = artist != null ? artist : "Unknown Artist";
        this.title = title != null ? title : "Unknown Title";
        this.album = album != null ? album : "Unknown Album";
        this.albumAuthor = albumAuthor != null ? albumAuthor : "Unknown Album Author";
        this.progress = progress;
        this.maxProgress = maxProgress;
        this.startTime = startTime != null ? startTime : LocalDateTime.now();
        this.endTime = endTime != null ? endTime : LocalDateTime.now();
        this.timeListened = timeListened;
    }

    public static SongData of(MediaController mediaController) {
        MediaMetadata metadata = mediaController.getMetadata();
        String artist = metadata != null ? metadata.getString(MediaMetadata.METADATA_KEY_ARTIST) : "Unknown Artist";
        String title = metadata != null ? metadata.getString(MediaMetadata.METADATA_KEY_TITLE) : "Unknown Title";
        String album = metadata != null ? metadata.getString(MediaMetadata.METADATA_KEY_ALBUM) : "Unknown Album";
        String albumAuthor = metadata != null ? metadata.getString(MediaMetadata.METADATA_KEY_ALBUM_ARTIST) : "Unknown Album Author";
        long progress = mediaController.getPlaybackState().getPosition();
        long maxProgress = metadata != null ? metadata.getLong(MediaMetadata.METADATA_KEY_DURATION) : 0;
        return new SongData(artist, title, album, albumAuthor, progress, maxProgress, LocalDateTime.now(), null, 0);
    }

    public String getIdentifier() {
        return artist + "," + title + "," + album;
    }

    public void incrementTimeListened() {
        this.timeListened++;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getArtist() {
        return artist;
    }

    public String getTitle() {
        return title;
    }

    public String getAlbum() {
        return album;
    }

    public String getAlbumAuthor() {
        return albumAuthor;
    }

    public long getProgress() {
        return progress;
    }

    public void setProgress(long progress) {
        this.progress = progress;
    }

    public long getMaxProgress() {
        return maxProgress;
    }

    public LocalDateTime getStartTime() {
        return startTime;
    }

    public LocalDateTime getEndTime() {
        return endTime;
    }

    public void setEndTime(LocalDateTime endTime) {
        this.endTime = endTime;
    }

    public int getTimeListened() {
        return timeListened;
    }

    public void setTimeListened(int timeListened) {
        this.timeListened = timeListened;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SongData songData = (SongData) o;
        return Objects.equals(artist, songData.artist) && Objects.equals(title, songData.title) && Objects.equals(album, songData.album) && Objects.equals(albumAuthor, songData.albumAuthor);
    }

    @Override
    public int hashCode() {
        return Objects.hash(artist, title, album, albumAuthor);
    }

    @Override
    public String toString() {
        return "SongData{" +
                "artist='" + artist + '\'' +
                ", title='" + title + '\'' +
                ", album='" + album + '\'' +
                ", albumAuthor='" + albumAuthor + '\'' +
                ", maxProgress=" + maxProgress +
                ", startTime=" + startTime +
                ", progress=" + progress +
                ", endTime=" + endTime +
                ", timeListened=" + timeListened +
                '}';
    }

}

File: lib/song_data.dart
class SongData {
  final int id;
  final String artist;
  final String title;
  final String album;
  final String albumAuthor;
  final int progress;
  final int maxProgress;
  final DateTime startTime;
  final DateTime endTime;
  final int timeListened;

  SongData({
    required this.id,
    required this.artist,
    required this.title,
    required this.album,
    required this.albumAuthor,
    required this.progress,
    required this.maxProgress,
    required this.startTime,
    required this.endTime,
    required this.timeListened,
  });

  factory SongData.fromJson(Map<String, dynamic> json) {
    return SongData(
      id: json['id'] ?? 0,
      artist: json['artist'] ?? '',
      title: json['title'] ?? '',
      album: json['album'] ?? '',
      albumAuthor: json['albumAuthor'] ?? '',
      progress: json['progress'] ?? 0,
      maxProgress: json['maxProgress'] ?? 0,
      startTime: DateTime.parse(json['startTime'] ?? '1970-01-01T00:00:00Z'),
      endTime: DateTime.parse(json['endTime'] ?? '1970-01-01T00:00:00Z'),
      timeListened: json['timeListened'] ?? 0,
    );
  }

  String getIdentifier() {
    return "$artist,$title,$album";
  }

  SongTileData toSongTileData() {
    return SongTileData(artist, title, album, albumAuthor, timeListened, 1);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SongData &&
          runtimeType == other.runtimeType &&
          artist == other.artist &&
          title == other.title &&
          album == other.album &&
          albumAuthor == other.albumAuthor;

  @override
  int get hashCode =>
      artist.hashCode ^ title.hashCode ^ album.hashCode ^ albumAuthor.hashCode;
}

class SongTileData {
  final String artist;
  final String title;
  final String album;
  final String albumAuthor;
  int allTimeListened;
  int listenCount;

  SongTileData(this.artist, this.title, this.album, this.albumAuthor,
      this.allTimeListened, this.listenCount);

  String getIdentifier() {
    return "$artist,$title,$album";
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SongTileData &&
          runtimeType == other.runtimeType &&
          artist == other.artist &&
          album == other.album &&
          albumAuthor == other.albumAuthor &&
          listenCount == other.listenCount;

  @override
  int get hashCode =>
      artist.hashCode ^
      album.hashCode ^
      albumAuthor.hashCode ^
      listenCount.hashCode;
}
